cli/src/main/kotlin/org/jetbrains/research/pynose/headless/HeadlessRunner.kt

```diff
@@ -0,0 +1,357 @@
+            }
+            gatherJsonFunctionInformation(inspectionName, holder, jsonFileResultArray)
+            gatherCsvInformation(unittest, inspectionName, holder, psiFile)
+        }
+    }
+
+    private fun initOutputJsonFile(outputDir: String, projectName: String): File {
+        val jsonOutputFileName = "$outputDir$separator${projectName}_ext_stats.json"
+        val jsonFile = File(jsonOutputFileName)
+        File(outputDir).mkdirs()
+        jsonFile.createNewFile()
+        return jsonFile
+    }
+
+    private fun writeToJsonFile(projectResult: JsonArray, jsonFile: File) {
+        val jsonString =
+            GsonBuilder().setPrettyPrinting().create().toJson(JsonParser.parseString(projectResult.toString()))
+        try {
+            jsonFile.writeText(jsonString)
+        } catch (e: IOException) {
+            e.printStackTrace()
+        }
+    }
+
+    private fun writeToCsvFile(outputDir: String, projectName: String, ut: Boolean) {
+        val sortedCsvMap = TreeMap(if (ut) unittestCsvMap else pytestCsvMap)
+        val csvOutputFileName =
+            "$outputDir${separator}${if (ut) "unittest" else "pytest"}${separator}${projectName}_stats.csv"
+        File("$outputDir${separator}${if (ut) "unittest" else "pytest"}").mkdirs()
+        File(csvOutputFileName).createNewFile()
+        val writer = Paths.get(csvOutputFileName).bufferedWriter()
+        val csvPrinter = CSVPrinter(writer, CSVFormat.DEFAULT)
+        val header = mutableListOf("project_name", "test_file_count")
+        val data = mutableListOf(projectName, { if (ut) unittestFileCount else pytestFileCount }.toString())
+
+        sortedCsvMap.keys.forEach { header.add(it) }
+        sortedCsvMap.keys.forEach { data.add(sortedCsvMap[it]?.size.toString()) }
+        if (!(if (ut) aggregatedUnittestHasHeader else aggregatedPytestHasHeader)) {
+            agCsvUnittestData.add(header)
+        }
+        if (ut) aggregatedUnittestHasHeader else aggregatedPytestHasHeader = true
+        (if (ut) agCsvUnittestData else agCsvPytestData).add(data)
+
+        csvPrinter.use {
+            it.printRecord(header)
+            it.printRecord(data)
+        }
+    }
+
+    private fun writeToAggregatedCsv(outputDir: String, testFramework: String) {
+        val agCsvOutputFileName = "$outputDir${separator}$testFramework${separator}aggregated_stats.csv"
+        val csvFile = File(agCsvOutputFileName)
+        File("$outputDir${separator}pytest").mkdirs()
+        csvFile.createNewFile()
+        val writer = Paths.get(agCsvOutputFileName).bufferedWriter()
+        val csvPrinter = CSVPrinter(writer, CSVFormat.DEFAULT)
+        if (testFramework == "pytest") {
+            agCsvPytestData.forEach { csvPrinter.printRecord(it) }
+        } else {
+            agCsvUnittestData.forEach { csvPrinter.printRecord(it) }
+        }
+        csvPrinter.flush()
+        csvPrinter.close()
+    }
+
+    override fun main(args: List<String>) {
```

Comment:
Public functions should be on the top of the file, otherwise it's harder to understand what's going on here at first look